govcd/vapp_network.go

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+	"net/http"
+)
+
+// UpdateNetworkFirewallRules updates vApp networks firewall rules.
+// Returns pointer to types.VAppNetwork or error
+func (vapp *VApp) UpdateNetworkFirewallRules(networkId string, firewallRules []*types.FirewallRule, defaultAction string, logDefaultAction bool) (*types.VAppNetwork, error) {
+	task, err := vapp.UpdateNetworkFirewallRulesAsync(networkId, firewallRules, defaultAction, logDefaultAction)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	return vapp.GetVappNetworkById(networkId, false)
+}
+
+// UpdateNetworkFirewallRulesAsync asynchronously updates vApp networks firewall rules.
+// Returns task or error
+func (vapp *VApp) UpdateNetworkFirewallRulesAsync(networkId string, firewallRules []*types.FirewallRule, defaultAction string, logDefaultAction bool) (Task, error) {
+	util.Logger.Printf("[TRACE] UpdateNetworkFirewallRulesAsync with values: id: %s and firewallServiceConfiguration: %#v", networkId, firewallRules)
+	uuid := extractUuid(networkId)
+	networkToUpdate, err := vapp.GetVappNetworkById(uuid, true)
+	if err != nil {
+		return Task{}, err
+	}
+
+	if networkToUpdate.Configuration.Features == nil {
+		networkToUpdate.Configuration.Features = &types.NetworkFeatures{}
+	}
+	networkToUpdate.Xmlns = types.XMLNamespaceVCloud
+
+	if networkToUpdate.Configuration.Features.FirewallService == nil {
+		return Task{}, fmt.Errorf("provided network isn't connecd to org network or isn't fenced")
+	}
+	networkToUpdate.Configuration.Features.FirewallService.LogDefaultAction = logDefaultAction
+	networkToUpdate.Configuration.Features.FirewallService.DefaultAction = defaultAction
+	networkToUpdate.Configuration.Features.FirewallService.FirewallRule = firewallRules
+
+	apiEndpoint := vapp.client.VCDHREF
+	apiEndpoint.Path += "/network/" + uuid
```

Comment:
I believe this deserves a comment, because one may think that `/network` is an Org VDC network. In fact, VCD API reference for GET on `/network` does actually return `OrgNetworkType`:

https://code.vmware.com/apis/722/doc/doc/operations/GET-Network.html
> Output parameters
>    OrgNetworkType 

Only when you dig more you find that the PUT on the same URL accepts `VAppNetworkType`:

https://code.vmware.com/apis/722/doc/doc/operations/PUT-Network.html
> Input type:
>    VAppNetworkType

Reply:
added